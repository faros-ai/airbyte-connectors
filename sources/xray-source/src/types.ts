interface Authentication {
  client_id: string;
  client_secret: string;
}

export interface XrayConfig {
  authentication: Authentication;
  projects: ReadonlyArray<string>;
  cutoff_days?: number;
  api_timeout?: number;
  api_page_limit?: number;
  api_max_retries?: number;
}
