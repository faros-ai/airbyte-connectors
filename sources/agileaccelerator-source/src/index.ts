import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {
  Agileaccelerator,
  AgileacceleratorConfig,
} from './agileaccelerator/agileaccelerator';
import {Works} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AgileacceleratorSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AgileAccelerator source implementation. */
export class AgileacceleratorSource extends AirbyteSourceBase<AgileacceleratorConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AgileacceleratorConfig
  ): Promise<[boolean, VError]> {
    try {
      const agileaccelerator = await Agileaccelerator.instance(
        config,
        this.logger
      );
      await agileaccelerator.checkConnection(config);
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AgileacceleratorConfig): AirbyteStreamBase[] {
    return [new Works(config, this.logger)];
  }
}
