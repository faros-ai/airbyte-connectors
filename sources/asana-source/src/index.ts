import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import {AirbyteSourceLogger} from 'faros-airbyte-cdk';
import VError from 'verror';

import {Asana, AsanaConfig} from './asana';
import {
  Projects,
  ProjectTasks,
  Tags,
  Tasks,
  Users,
  Workspaces,
} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new AsanaSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** Asana source implementation. */
export class AsanaSource extends AirbyteSourceBase<AsanaConfig> {
  get type(): string {
    return 'asana';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: AsanaConfig): Promise<[boolean, VError]> {
    try {
      const asana = Asana.instance(config, this.logger);
      await asana.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AsanaConfig): AirbyteStreamBase[] {
    return [
      new Projects(config, this.logger),
      new Tags(config, this.logger),
      new Tasks(config, this.logger),
      new Users(config, this.logger),
      new Workspaces(config, this.logger),
      new ProjectTasks(config, this.logger),
    ];
  }
}
