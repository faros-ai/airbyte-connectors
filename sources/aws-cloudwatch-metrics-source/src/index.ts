import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceLogger,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {CloudWatch, Config} from './cloudwatch';
import {Metrics} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new CloudWatchMetricsSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** CloudWatch Metrics source implementation. */
export class CloudWatchMetricsSource extends AirbyteSourceBase<Config> {
  get type(): string {
    return 'cloudwatch-metrics';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: Config): Promise<[boolean, VError]> {
    try {
      const cloudWatch = CloudWatch.instance(config);
      await cloudWatch.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: Config): AirbyteStreamBase[] {
    return [new Metrics(config, this.logger)];
  }
}
