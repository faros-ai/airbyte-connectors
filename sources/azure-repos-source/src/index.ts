import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {AzureRepoConfig, AzureRepos} from './azure-repos';
import {PullRequests, Repositories, Users} from './streams';
import {Commits} from './streams/commits';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AzureRepoSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AzureRepo source implementation. */
export class AzureRepoSource extends AirbyteSourceBase<AzureRepoConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: AzureRepoConfig): Promise<[boolean, VError]> {
    try {
      const azureRepos = await AzureRepos.make(config, this.logger);
      await azureRepos.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AzureRepoConfig): AirbyteStreamBase[] {
    return [
      new Commits(config, this.logger),
      new PullRequests(config, this.logger),
      new Repositories(config, this.logger),
      new Users(config, this.logger),
    ];
  }
}
