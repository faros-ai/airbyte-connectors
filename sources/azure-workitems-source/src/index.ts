import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {AzureWorkitems, AzureWorkitemsConfig} from './azure-workitems';
import {Boards, Iterations, Users} from './streams';
import {Workitems} from './streams/workitems';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AzureWorkitemsSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** Example source implementation. */
export class AzureWorkitemsSource extends AirbyteSourceBase<AzureWorkitemsConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AzureWorkitemsConfig
  ): Promise<[boolean, VError]> {
    try {
      const azureActiveDirectory = await AzureWorkitems.instance(config);
      await azureActiveDirectory.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AzureWorkitemsConfig): AirbyteStreamBase[] {
    return [
      new Workitems(config, this.logger),
      new Users(config, this.logger),
      new Iterations(config, this.logger),
      new Boards(config, this.logger),
    ];
  }
}
