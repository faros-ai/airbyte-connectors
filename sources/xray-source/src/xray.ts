import axios, {AxiosInstance} from 'axios';
import createAuthRefreshInterceptor from 'axios-auth-refresh';
import {AirbyteLogger} from 'faros-airbyte-cdk';
import {
  Test,
  TestExecution,
  TestKey,
  TestPlan,
  TestRun,
} from 'faros-airbyte-common/xray';
import {makeAxiosInstanceWithRetry, wrapApiError} from 'faros-js-client';
import * as fs from 'fs';
import {get} from 'lodash';
import path from 'path';
import {Memoize} from 'typescript-memoize';
import VError, {MultiError} from 'verror';

import {XrayConfig} from './types';

const XRAY_CLOUD_BASE_URL = 'https://xray.cloud.getxray.app/api/v2';
const XRAY_API_DEFAULT_TIMEOUT = 0;
const XRAY_PAGE_LIMIT = 100;

export class Xray {
  private static xray: Xray;
  constructor(
    private readonly api: AxiosInstance,
    private readonly limit,
    private readonly logger?: AirbyteLogger
  ) {}

  static async instance(
    config: XrayConfig,
    logger?: AirbyteLogger
  ): Promise<Xray> {
    if (Xray.xray) return Xray.xray;
    const api = makeAxiosInstanceWithRetry(
      {
        baseURL: XRAY_CLOUD_BASE_URL,
        timeout: config.api_timeout ?? XRAY_API_DEFAULT_TIMEOUT,
        headers: {
          'Content-Type': 'application/json',
        },
        maxContentLength: Infinity, //default is 2000 bytes
      },
      logger?.asPino(),
      config.api_max_retries
    );

    const auth = config.authentication;
    if (!auth?.client_id || !auth?.client_secret) {
      throw new VError(
        'Please provide Xray Cloud authentication details, ' +
          'Client Id and a Client Secret'
      );
    }

    const refreshToken = async (): Promise<void> => {
      const token = await Xray.sessionToken(auth.client_id, auth.client_secret);
      api.defaults.headers.common['Authorization'] = `Bearer ${token}`;
    };
    createAuthRefreshInterceptor(api, refreshToken, {statusCodes: [401]});

    // Limit should be a max of 100
    // https://docs.getxray.app/display/XRAYCLOUD/GraphQL+API
    const limit = config.api_page_limit
      ? Math.min(config.api_page_limit, XRAY_PAGE_LIMIT)
      : XRAY_PAGE_LIMIT;
    return new Xray(api, limit, logger);
  }

  private static async sessionToken(
    clientId: string,
    clientSecret: string
  ): Promise<string> {
    try {
      const {data} = await axios.post(`${XRAY_CLOUD_BASE_URL}/authenticate`, {
        client_id: clientId,
        client_secret: clientSecret,
      });
      return data;
    } catch (err: any) {
      throw wrapApiError(err, 'failed to get authentication token');
    }
  }

  private static readQueryFile(fileName: string): string {
    return fs.readFileSync(
      path.join(__dirname, '..', 'resources', 'queries', fileName),
      'utf8'
    );
  }

  private async *paginate(
    queryFile: string,
    resultKey: string,
    variables?: any
  ): AsyncGenerator<any> {
    const query = Xray.readQueryFile(queryFile);
    let hasNextPage = true;
    let start = null;
    let count = 0;

    while (hasNextPage) {
      let response;
      try {
        response = await this.api.post('/graphql', {
          query,
          variables: {
            limit: this.limit,
            start,
            ...variables,
          },
        });
      } catch (err: any) {
        throw wrapApiError(err, 'failed to fetch data');
      }
      if (response.data?.errors) {
        throw new MultiError(
          response.data.errors.map((e: any) => new VError(e.message))
        );
      }

      const data = get(response.data?.data, resultKey);
      const results = data.results;
      for (const result of results) {
        yield result;
        count++;
      }
      hasNextPage = data.total != count && data.results.length === data.limit;
      start = results.length;
    }
  }

  async checkConnection(): Promise<void> {
    // Basic query to check if the connection is working
    const query = `
      query getType {
        __type(name: "Test") {
          name
          kind
          description
        }
      }
    `;
    await this.api.post('/graphql', {query});
  }

  @Memoize()
  async getTestPlans(project: string): Promise<ReadonlyArray<TestPlan>> {
    const varibles = {jql: `project = ${project}`};
    const plans = [];
    for await (const plan of this.paginate(
      'get-test-plans.gql',
      'getTestPlans',
      varibles
    )) {
      const {key, summary, description, labels} = plan.jira;
      plans.push({
        issueId: plan.issueId,
        key,
        summary,
        description,
        labels,
      });
    }
    return plans;
  }

  async *getTests(
    project: string,
    modifiedSince: string
  ): AsyncGenerator<Test> {
    const variables = {jql: `project = ${project}`, modifiedSince};
    for await (const test of this.paginate(
      'get-tests.gql',
      'getTests',
      variables
    )) {
      const {key, summary, description, labels} = test.jira;
      yield {
        issueId: test.issueId,
        key,
        summary,
        description,
        labels,
        gherkin: test.gherkin,
        unstructured: test.unstructured,
        testType: test.testType,
        status: test.status,
        steps: test.steps,
        preconditions: test.preconditions?.results?.map((p: any) => {
          const {key, rank} = p.jira;
          return {
            issueId: p.issueId,
            key,
            definition: p.definition,
            rank,
            preconditionType: p.preconditionType,
          };
        }),
        project,
        lastModified: test.lastModified,
      };
    }
  }

  async *getTestPlanTests(planId: string): AsyncGenerator<TestKey> {
    for await (const test of this.paginate(
      'get-test-plan-tests.gql',
      'getTestPlan.tests',
      {issueId: planId}
    )) {
      yield {
        issueId: test.issueId,
        key: test.jira.key,
      };
    }
  }

  async *getTestRuns(modifiedSince: string): AsyncGenerator<TestRun> {
    const variables = {modifiedSince};
    for await (const run of this.paginate(
      'get-test-runs.gql',
      'getTestRuns',
      variables
    )) {
      yield {
        id: run.id,
        startedOn: run.startedOn,
        finishedOn: run.finishedOn,
        defects: run.defects,
        status: run.status,
        steps: run.steps,
        lastModified: run.lastModified,
        test: {
          issueId: run.test.issueId,
          key: run.test.jira.key,
        },
        testVersion: run.testVersion,
        testExecution: {
          issueId: run.testExecution.issueId,
          key: run.testExecution.jira.key,
        },
      };
    }
  }

  async *getTestExecutions(
    project: string,
    modifiedSince: string
  ): AsyncGenerator<TestExecution> {
    const variables = {jql: `project = ${project}`, modifiedSince};
    for await (const execution of this.paginate(
      'get-test-executions.gql',
      'getTestExecutions',
      variables
    )) {
      const {key, summary, description, labels} = execution.jira;
      yield {
        issueId: execution.issueId,
        key,
        summary,
        description,
        labels,
        testEnvironments: execution.testEnvironments,
        lastModified: execution.lastModified,
        project,
      };
    }
  }
}
