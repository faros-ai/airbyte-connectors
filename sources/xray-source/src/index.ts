import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceLogger,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {TestExecutions} from './streams/testExecutions';
import {TestPlans} from './streams/testPlans';
import {TestPlanTests} from './streams/testPlanTests';
import {TestRuns} from './streams/testRuns';
import {Tests} from './streams/tests';
import {XrayConfig} from './types';
import {Xray} from './xray';

export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new XraySource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

export class XraySource extends AirbyteSourceBase<XrayConfig> {
  get type(): string {
    return 'xray';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }

  async checkConnection(config: XrayConfig): Promise<[boolean, VError]> {
    try {
      const xray = await Xray.instance(config, this.logger);
      await xray.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }

  streams(config: XrayConfig): AirbyteStreamBase[] {
    return [
      new TestPlans(config, this.logger),
      new Tests(config, this.logger),
      new TestPlanTests(config, this.logger),
      new TestExecutions(config, this.logger),
      new TestRuns(config, this.logger),
    ];
  }
}
