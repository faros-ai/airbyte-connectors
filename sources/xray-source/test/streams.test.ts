import {
  AirbyteLogLevel,
  AirbyteSourceLogger,
  SyncMode,
} from 'faros-airbyte-cdk';
import fs from 'fs-extra';

import * as sut from '../src/index';
import {Xray} from '../src/xray';

function readTestResourceFile(fileName: string): any {
  return JSON.parse(fs.readFileSync(`test_files/${fileName}`, 'utf8'));
}

describe('streams', () => {
  const logger = new AirbyteSourceLogger(
    // Shush messages in tests, unless in debug
    process.env.LOG_LEVEL === 'debug'
      ? AirbyteLogLevel.DEBUG
      : AirbyteLogLevel.FATAL
  );

  const config = {
    authentication: {client_id: 'client_id', client_secret: 'client_secret'},
    projects: ['TEST'],
  };

  async function testStream(
    streamIndex: number,
    responseFileOrFn: string | jest.Mock
  ): Promise<void> {
    const postFn =
      typeof responseFileOrFn === 'string'
        ? jest.fn().mockResolvedValue({
            data: {data: readTestResourceFile(responseFileOrFn)},
          })
        : responseFileOrFn;

    Xray.instance = jest.fn().mockImplementation(() => {
      return new Xray({post: postFn} as any, logger);
    });

    const source = new sut.XraySource(logger);
    const streams = source.streams(config);
    const stream = streams[streamIndex];
    const iter = stream.readRecords(SyncMode.FULL_REFRESH, undefined, {
      project: 'TEST',
    });

    const items = [];
    for await (const item of iter) {
      items.push(item);
    }
    expect(items).toMatchSnapshot();
  }

  test('test plans', async () => {
    await testStream(0, 'getPlans.json');
  });

  test('tests', async () => {
    await testStream(1, 'getTests.json');
  });

  test('test plan tests', async () => {
    const mockFn = jest
      .fn()
      .mockResolvedValueOnce({
        data: {data: readTestResourceFile('getPlans.json')},
      })
      .mockResolvedValue({
        data: {data: readTestResourceFile('getTestPlanTests.json')},
      });
    await testStream(2, mockFn);
  });

  test('test executions', async () => {
    await testStream(3, 'getTestExecutions.json');
  });

  test('test runs', async () => {
    await testStream(4, 'getTestRuns.json');
  });
});
