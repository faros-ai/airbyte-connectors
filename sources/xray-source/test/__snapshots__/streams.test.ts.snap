// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`streams test executions 1`] = `
[
  {
    "description": "Test Execution Desc",
    "issueId": "23133",
    "key": "TEST-191",
    "labels": [],
    "lastModified": "2024-06-05T15:17:34.780Z",
    "project": "TEST",
    "summary": "Test Execution",
    "testEnvironments": [
      "Dev",
    ],
  },
]
`;

exports[`streams test plan tests 1`] = `
[
  {
    "planIssueId": "23082",
    "planKey": "TEST-11692",
    "testIssueId": "23096",
    "testKey": "TEST-187",
  },
  {
    "planIssueId": "23082",
    "planKey": "TEST-11692",
    "testIssueId": "23097",
    "testKey": "TEST-188",
  },
]
`;

exports[`streams test plans 1`] = `
[
  {
    "description": "Example test plan",
    "issueId": "23082",
    "key": "TEST-11692",
    "labels": [],
    "summary": "Test Plan - Example1",
  },
]
`;

exports[`streams test runs 1`] = `
[
  {
    "defects": [],
    "finishedOn": "2024-06-05T14:14:51.071Z",
    "id": "6660728b897a8ced80c26f0a",
    "lastModified": "2024-06-05T14:14:51.075Z",
    "startedOn": "2024-06-05T14:14:51.071Z",
    "status": {
      "name": "PASSED",
    },
    "steps": [],
    "test": {
      "issueId": "23097",
      "key": "TEST-188",
    },
    "testExecution": {
      "issueId": "23133",
      "key": "TEST-191",
    },
    "testVersion": {
      "name": "v1",
    },
  },
  {
    "defects": [
      "23135",
    ],
    "finishedOn": "2024-06-05T14:59:49.234Z",
    "id": "6660738d897a8ced80c5db6f",
    "lastModified": "2024-06-05T14:59:49.241Z",
    "startedOn": "2024-06-05T14:59:49.234Z",
    "status": {
      "name": "FAILED",
    },
    "steps": [],
    "test": {
      "issueId": "23096",
      "key": "TEST-187",
    },
    "testExecution": {
      "issueId": "23133",
      "key": "TEST-191",
    },
    "testVersion": {
      "name": "v1",
    },
  },
  {
    "defects": [],
    "finishedOn": "2024-06-05T14:40:23.960Z",
    "id": "6660743e897a8ced80c933e3",
    "lastModified": "2024-06-05T14:41:33.995Z",
    "startedOn": "2024-06-05T14:37:01.124Z",
    "status": {
      "name": "PASSED",
    },
    "steps": [
      {
        "id": "8865a132-8b37-4eb3-bb46-f03cd32adf82",
        "result": null,
        "status": {
          "name": "PASSED",
        },
      },
      {
        "id": "c114757d-7c5b-4983-87c0-a88f2a1d8281",
        "result": null,
        "status": {
          "name": "PASSED",
        },
      },
    ],
    "test": {
      "issueId": "23134",
      "key": "TEST-192",
    },
    "testExecution": {
      "issueId": "23133",
      "key": "TEST-191",
    },
    "testVersion": {
      "name": "v1",
    },
  },
]
`;

exports[`streams tests 1`] = `
[
  {
    "description": "Third test",
    "gherkin": null,
    "issueId": "23098",
    "key": "TEST-189",
    "labels": [],
    "lastModified": "2024-06-04T02:10:00.825Z",
    "preconditions": [],
    "project": "TEST",
    "status": {
      "name": "TO DO",
    },
    "steps": [
      {
        "action": "Do something test 4",
        "data": "",
        "id": "8865a132-8b37-4eb3-bb46-f03cd32adf82",
        "result": "",
      },
      {
        "action": "Something else to do",
        "data": "",
        "id": "c114757d-7c5b-4983-87c0-a88f2a1d8281",
        "result": "",
      },
    ],
    "summary": "Example test 3",
    "testType": {
      "kind": "Steps",
      "name": "Manual",
    },
    "unstructured": null,
  },
  {
    "description": "example test 2",
    "gherkin": null,
    "issueId": "23097",
    "key": "TEST-188",
    "labels": [],
    "lastModified": "2024-06-04T18:40:36.017Z",
    "preconditions": [
      {
        "definition": null,
        "issueId": "23105",
        "key": "TEST-190",
        "preconditionType": {
          "kind": "Steps",
          "name": "Manual",
        },
        "rank": 1,
      },
    ],
    "project": "TEST",
    "status": {
      "name": "TO DO",
    },
    "steps": [],
    "summary": "Example test 2",
    "testType": {
      "kind": "Steps",
      "name": "Manual",
    },
    "unstructured": null,
  },
  {
    "description": "First test example",
    "gherkin": null,
    "issueId": "23096",
    "key": "TEST-187",
    "labels": [
      "label1",
      "label2",
    ],
    "lastModified": "2024-06-05T14:20:07.166Z",
    "preconditions": [],
    "project": "TEST",
    "status": {
      "name": "TO DO",
    },
    "steps": [],
    "summary": "Example test 1",
    "testType": {
      "kind": "Steps",
      "name": "Manual",
    },
    "unstructured": null,
  },
]
`;
